### PR TITLE
Remove allocations related to views

### PR DESCRIPTION
**Pull request by amontoison**

### CODE REVIEW

File: src/moi_nls_model.jl [hunk 1]
```diff
@@ -93,7 +93,7 @@ function NLPModels.residual!(nls::MathOptNLSModel, x::AbstractVector, Fx::Abstra
       x,
       view(Fx, nls.nls_meta.lin),
     )
-    Fx[nls.nls_meta.lin] .+= nls.linequ.constants
+    view(Fx, nls.nls_meta.lin) .+= nls.linequ.constants
```
Comment (tmigot) [inline src/moi_nls_model.jl:96]: Would it be cleaner to use `view` like this everywhere instead of loops?

**Review by tmigot — CHANGES_REQUESTED** [review]
This looks a great improvement! I just have cosmetic comments.

File: src/moi_nls_model.jl (hunk 1 above)
Comment (amontoison) [reply src/moi_nls_model.jl:96]: Yes, good idea!
I also added a function `coo_sym_add_mul!`.

File: src/moi_nls_model.jl [hunk 2]
```diff
@@ -376,17 +366,16 @@ function NLPModels.hess_structure!(
   cols::AbstractVector{<:Integer},
 )
   if nls.nls_meta.nlin > 0
-    for index = 1:(nls.lls.nnzh)
-      rows[index] = nls.lls.hessian.rows[index]
-      cols[index] = nls.lls.hessian.cols[index]
-    end
+    view(rows, 1:nls.lls.nnzh) .= nls.lls.hessian.rows
```
Comment (tmigot) [inline src/moi_nls_model.jl:369]: Is there no coverage here because we have no NLS with linear constraints?

File: src/utils.jl [hunk 3]
```diff
@@ -71,10 +71,35 @@ function replace!(ex, x)
   end
 end
 
+"""
+    coo_sym_add_mul!(rows, cols, vals, x, y, α)
+
+Update of the form `y ← y + αAx` where `A` is a symmetric matrix given by `(rows, cols, vals)`.
+Only one triangle of `A` should be passed.
+"""
+function coo_sym_add_mul!(
```
Comment (tmigot) [inline src/utils.jl:80]: This function looks a bit like the function `coo_sym_mul!` in `SparseMatricesCOO.jl` https://github.com/JuliaSmoothOptimizers/SparseMatricesCOO.jl/blob/6bfc6ca2cdcc6fd746aebb535425754993578dcc/src/coo_linalg.jl#L66 #109

Comment (tmigot) [reply src/utils.jl:80]: If that's the same function, we could reuse your docstring :)

File: src/moi_nls_model.jl [hunk 4]
```diff
@@ -376,17 +366,16 @@ function NLPModels.hess_structure!(
...
+    view(cols, 1:nls.lls.nnzh) .= nls.lls.hessian.cols
   end
   if nls.nls_meta.nnln > 0
     hesslag_struct = MOI.hessian_lagrangian_structure(nls.ceval)
     for index = (nls.lls.nnzh + 1):(nls.meta.nnzh)
       shift_index = index - nls.lls.nnzh
-      rows[index] = hesslag_struct[shift_index][1]
-      cols[index] = hesslag_struct[shift_index][2]
+      row, col = hesslag_struct[shift_index]
```
Comment (tmigot) [inline src/moi_nls_model.jl:376]: Can you check if this allocates? Sometimes, it does when we create a tuple.

**Review by tmigot — CHANGES_REQUESTED** [review]
Thanks @amontoison ! That's just minor comments.

File: src/moi_nls_model.jl (hunk 2 above)
Comment (tmigot) [reply src/moi_nls_model.jl:369]: We can handle this in a separate issue.

**Review by tmigot — APPROVED** [review]
Thanks

**Pull request merged**